pkg/kev/render.go

```diff
@@ -0,0 +1,183 @@
+
+	"github.com/appvia/kev/pkg/kev/config"
+	"github.com/appvia/kev/pkg/kev/converter"
+	kmd "github.com/appvia/komando"
+	"github.com/pkg/errors"
+)
+
+func NewRenderRunner(workingDir string, opts ...Options) *RenderRunner {
+	runner := &RenderRunner{Project: &Project{workingDir: workingDir}}
+	runner.Init(opts...)
+	return runner
+}
+
+// Run executes the runner returning results that can be written to disk
+func (r *RenderRunner) Run() (map[string]string, error) {
+	if err := r.LoadProject(); err != nil {
+		return nil, err
+	}
+
+	if err := r.ValidateSources(r.manifest.Sources, config.SecretMatchers); err != nil {
+		return nil, err
+	}
+
+	if err := r.ValidateSkaffoldIfRequired(); err != nil {
+		return nil, err
+	}
+
+	if err := r.ValidateEnvSources(config.SecretMatchers); err != nil {
+		return nil, err
+	}
+
+	if err := r.ReconcileEnvsAndWriteUpdates(); err != nil {
+		return nil, err
+	}
+
+	return r.RenderManifests()
+}
+
+func (r *RenderRunner) LoadProject() error {
+	sg := r.UI.StepGroup()
+	defer sg.Done()
+
+	if !ManifestExistsForPath(path.Join(r.workingDir, ManifestFilename)) {
+		err := errors.Errorf("Missing project manifest: %s", ManifestFilename)
+		renderStepError(r.UI, sg.Add(""), renderStepLoad, err)
+		return err
+	}
+
+	manifest, err := LoadManifest(r.workingDir)
+	if err != nil {
+		renderStepError(r.UI, sg.Add(""), renderStepLoad, err)
+		return err
+	}
+	r.manifest = manifest
+	r.manifest.UI = r.UI
+	return nil
+}
+
+func (r *RenderRunner) ValidateSkaffoldIfRequired() error {
+	if len(r.manifest.Skaffold) == 0 {
+		return nil
+	}
+
+	r.UI.Header("Verifying Skaffold...")
+	sg := r.UI.StepGroup()
+	defer sg.Done()
+
+	step := sg.Add("Ensuring Skaffold manifest available")
+	if !ManifestExistsForPath(r.manifest.Skaffold) {
+		err := errors.Errorf("Missing Skaffold manifest %s", r.manifest.Skaffold)
+		renderStepError(r.UI, step, renderStepLoadSkaffold, err)
+		return err
+	}
+	step.Success("Skaffold manifest is available")
+	return nil
+}
+
+func (r *RenderRunner) ValidateEnvSources(matchers []map[string]string) error {
+	r.UI.Header("Validating deployment environments...")
+	var detectHit bool
+
+	filteredEnvs, err := r.manifest.GetEnvironments(r.config.envs)
+	if err != nil {
+		return err
+	}
+
+	for _, env := range filteredEnvs {
+		secretsDetected, err := r.detectSecretsInSources(env.ToSources(), matchers)
+		if err != nil {
+			return err
+		}
+		if secretsDetected {
+			detectHit = true
+		}
+	}
+
+	r.UI.Output("")
+	r.UI.Output("Validation successful!")
+	if detectHit {
+		r.UI.Output(fmt.Sprintf(`However, to prevent secrets leaking, see help page:
+%s`, SecretsReferenceUrl))
+	}
+
+	return nil
+}
+
+func (r *RenderRunner) ReconcileEnvsAndWriteUpdates() error {
+	r.UI.Header("Detecting project updates...")
+
+	if _, err := r.manifest.ReconcileConfig(r.config.envs...); err != nil {
+		return err
+	}
+
+	return r.manifest.Environments.Write()
+}
+
+func (r *RenderRunner) RenderManifests() (map[string]string, error) {
```

Comment:
Would be ideal to leave one line comments for exported methods.

Reply:
We use word `manifest` in quite a few places when referring to Kev manifest, Skaffold manifest and to Kubernetes manifests. Would be great to rename this method to make it more explicit as to what manifests we're talking about.

It'd be beneficial to make that distinction and stick to nomenclature in other places for clarity.